FWCore/Framework/src/EventSetupRecordIOVQueue.cc

```diff
@@ -0,0 +1,132 @@
+// -*- C++ -*-
+//
+// Package:     Framework
+// Class  :     EventSetupRecordIOVQueue
+//
+//
+// Author:      W. David Dagenhart
+// Created:     22 Feb 2019
+
+#include "FWCore/Framework/src/EventSetupRecordIOVQueue.h"
+#include "FWCore/Framework/interface/EventSetupRecordProvider.h"
+#include "FWCore/Concurrency/interface/WaitingTask.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include <exception>
+
+namespace edm {
+  namespace eventsetup {
+
+    EventSetupRecordIOVQueue::EventSetupRecordIOVQueue(unsigned int nConcurrentIOVs)
+        : nConcurrentIOVs_(nConcurrentIOVs), iovQueue_(nConcurrentIOVs), cacheIdentifier_(1) {
+      for (unsigned int i = 0; i < nConcurrentIOVs; ++i) {
+        isAvailable_.push_back(std::make_unique<std::atomic<bool>>(true));
+      }
+      waitForIOVsInFlight_ = edm::make_empty_waiting_task();
+      waitForIOVsInFlight_->increment_ref_count();
+      waitForIOVsInFlight_->increment_ref_count();
+    }
+
+    EventSetupRecordIOVQueue::~EventSetupRecordIOVQueue() {
+      { WaitingTaskHolder tmp{std::move(endIOVTaskHolder_)}; }
+      waitForIOVsInFlight_->decrement_ref_count();
+      waitForIOVsInFlight_->wait_for_all();
+    }
+
+    void EventSetupRecordIOVQueue::setNewIntervalForAnySubProcess() {
+      bool newIntervalForAnySubProcess = false;
+      for (auto const& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() == EventSetupRecordProvider::IntervalStatus::NewInterval) {
+          newIntervalForAnySubProcess = true;
+          break;
+        }
+      }
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->setNewIntervalForAnySubProcess(newIntervalForAnySubProcess);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::checkForNewIOVs(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                   WaitingTaskList& endIOVWaitingTasks,
+                                                   bool newEventSetupImpl) {
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->newIntervalForAnySubProcess()) {
+          startNewIOVAsync(taskToStartAfterIOVInit, endIOVWaitingTasks);
+          return;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() != EventSetupRecordProvider::IntervalStatus::Invalid) {
+          endIOVWaitingTasks.add(endIOVWaitingTask_);
+          break;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->continueIOV(newEventSetupImpl);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::startNewIOVAsync(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                    WaitingTaskList& endIOVWaitingTasks) {
+      ++cacheIdentifier_;
```

Comment:
Is there something guaranteeing that this function can only be called by one thread at a time? Else `cacheIdentifier_` modifications are not safe.

Reply:
I think the order of the calls is guaranteed. cacheIdentifier_ is incremented as part of the call to eventSetupForInstance (which calls checkForNewIOVs, which then calls startNewIOVAsync) and only if we know a new IOV must be started for that record before we can go forward. This same function creates the lambda which actually starts the IOV. The creation of the lambda occurs after the increment on the same thread. We cannot run the lambda until after its creation is complete. Then we use cacheIdentifier_ while running the lambda. Finally, we cannot run eventSetupForInstance again until after we have started the IOV, which allows starting the Lumi, which allow reading the next thing in the file and calling beginLumiAsync again if the next luminosity block is seen in the input. This is a bit complicated, but I think it works. Let me know if I've not completely convinced you as this is important if I've got it wrong.